From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: way-zer <himc.wicp@gmail.com>
Date: Wed, 20 Mar 2024 23:51:15 +0800
Subject: [PATCH] HC: ReplayController
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

兼容ARC的mrep格式
way-zer <himc.wicp@gmail.com> on 2024/4/22 at 18:10

实现功能按钮
way-zer <himc.wicp@gmail.com> on 2024/4/22 at 20:15
---
 core/src/mindustry/core/NetClient.java | 3 +++
 core/src/mindustry/io/TypeIO.java      | 2 +-
 core/src/mindustry/net/Net.java        | 2 ++
 3 files changed, 6 insertions(+), 1 deletion(-)

diff --git a/core/src/mindustry/core/NetClient.java b/core/src/mindustry/core/NetClient.java
index cd80be66ce1459e9276f71edb98ef8afe81ff6d9..74fdf93ba8907d17cb6ee1bd3e7e05708e883a2b 100644
--- a/core/src/mindustry/core/NetClient.java
+++ b/core/src/mindustry/core/NetClient.java
@@ -27,6 +27,7 @@ import mindustry.net.Packets.*;
 import mindustry.world.*;
 import mindustry.world.modules.*;
 import mindustryX.*;
+import mindustryX.features.*;
 
 import java.io.*;
 import java.util.*;
@@ -113,12 +114,14 @@ public class NetClient implements ApplicationListener{
                 return;
             }
 
+            ReplayController.onConnect(packet.addressTCP);
             net.send(c, true);
         });
 
         net.handleClient(Disconnect.class, packet -> {
             if(quietReset) return;
 
+            ReplayController.stop();
             connecting = false;
             logic.reset();
             platform.updateRPC();
diff --git a/core/src/mindustry/io/TypeIO.java b/core/src/mindustry/io/TypeIO.java
index c93d0921230fcbaacdd8cd673cf12fd4c313e823..93332c435d9383d9fb584e81f676171318f77c98 100644
--- a/core/src/mindustry/io/TypeIO.java
+++ b/core/src/mindustry/io/TypeIO.java
@@ -316,7 +316,7 @@ public class TypeIO{
             netClient.addRemovedEntity(entity.id());
         }
 
-        return null; //no need to actually return anything
+        return new UnitSyncContainer(entity); //MDTX: We should return non-null
     }
 
     public static void writeUnit(Writes write, Unit unit){
diff --git a/core/src/mindustry/net/Net.java b/core/src/mindustry/net/Net.java
index 7072e96706e4134bccc29c2cd6b40e2290b1072c..8929ef9d867729c779f809f08bb4bd075e7b9593 100644
--- a/core/src/mindustry/net/Net.java
+++ b/core/src/mindustry/net/Net.java
@@ -11,6 +11,7 @@ import mindustry.gen.*;
 import mindustry.net.Packets.*;
 import mindustry.net.Streamable.*;
 import mindustryX.events.*;
+import mindustryX.features.*;
 import net.jpountz.lz4.*;
 
 import java.io.*;
@@ -271,6 +272,7 @@ public class Net{
      */
     public void handleClientReceived(Packet object){
         object.handled();
+        ReplayController.onClientPacket(object);
 
         if(object instanceof StreamBegin b){
             streams.put(b.id, currentStream = new StreamBuilder(b));
